zinc-k8s-operator/src/charm.py

```diff
@@ -0,0 +1,145 @@
+        # The OCI image for Zinc only has two binaries: zincsearch and go-runner.
+        # We use go-runner to achieve the equivalent of `bash -c '/bin/zincsearch | tee PATH'`.
+        command = "/bin/go-runner --log-file=/var/lib/zincsearch/zinc.log --also-stdout=true --redirect-stderr=true /bin/zincsearch"
+        # Define a pebble layer and add it to the workload container.
+        layer: ops.pebble.LayerDict = {
+            "summary": "Zinc service",
+            "description": "Pebble layer that specifies how to run Zinc",
+            "services": {
+                "zinc": {
+                    "override": "replace",
+                    "summary": "zinc",
+                    "command": command,
+                    "startup": "enabled",
+                    "environment": {
+                        "ZINC_DATA_PATH": "/var/lib/zincsearch",
+                        "ZINC_FIRST_ADMIN_USER": self._zinc_config["admin_user"],
+                        "ZINC_FIRST_ADMIN_PASSWORD": self._zinc_config["admin_password"],
+                    },
+                },
+            },
+            "checks": {
+                "zinc-ready": {
+                    "override": "replace",
+                    "level": "ready",
+                    "period": "30s",
+                    "timeout": "5s",
+                    "threshold": 1,
+                    "http": {
+                        "url": f"{self._zinc.base_url}/version",
+                    },
+                }
+            },
+        }
+        self._pebble.add_layer("zinc", layer, combine=True)
+
+    def _on_get_admin_password(self, event: ops.ActionEvent):
+        """Return the initial admin password as an action response."""
+        if not self._update_zinc_config_from_peer_relation():
+            event.fail("admin password is not available")
+            return
+        event.set_results({
+            "admin-password": self._zinc_config["admin_password"]
+        })
+
+    def _update_zinc_config_from_peer_relation(self) -> bool:
+        """Fetch the initial admin password from the peer relation."""
+        relation = self.model.get_relation("zinc-peers")
+        if not relation:
+            return False # Relation not ready
+        secret_id = relation.data[self.app].get("admin-password")
+        if not secret_id:
+            return False # Secret not available
+        secret = self.model.get_secret(id=secret_id)
+        self._zinc_config["admin_password"] = secret.peek_content()["password"]
+        return True
+
+    def _update_peer_relation_from_zinc_config(self) -> bool:
+        """Store the initial admin password in the peer relation."""
+        relation = self.model.get_relation("zinc-peers")
+        if not relation:
+            return False # Relation not ready
+        secret = self.app.add_secret({
+            "password": self._zinc_config["admin_password"]
+        })
+        relation.data[self.app]["admin-password"] = secret.id
```

Comment:
Are we quitely relying on a Juju hook tool (?) feature where setting the same secret to the same value multiple times does not create new secret revisions?
If so, please document, it's surely not obvious to a new charmer.
If not, I'd rather see this method idempotent.